doc: bsim: Some more touch-ups

jori-nordic opened the pull request:
See commit messages.

jori-nordic committed:
doc: bsim: Add paragraph about runner conventions

Alberto had a bad time.

Signed-off-by: Jonathan Rico <jonathan.rico@nordicsemi.no>

jori-nordic committed:
doc: bsim: Reword test script conventions

Try to make it more terse, and remove duplicate lines.

Signed-off-by: Jonathan Rico <jonathan.rico@nordicsemi.no>

aescolar commented on doc/develop/test/bsim.rst:
```diff
@@ -132,6 +132,14 @@ and then directly run one of the tests:
+Characters
+----------
+
+Please don't use non-ASCII characters, like unicode or emoji in the scripts or test code.
```
I agree with this, but it seems a bit out of place to have guidelines like this here.
(This should apply to the whole tree, not just to tests)

jori-nordic replied:
will remove. I saw fabio is trying to add it to the compliance step anyways

aescolar commented on doc/develop/test/bsim.rst:
```diff
@@ -150,23 +158,33 @@ required, and some Zephyr users do not use or have west, but still use the build
+build system or compatibility changes if the tests are run in the same manner, with the same
+variables, etc..
+
+If you have a good idea for improving your test script, please make a PR changing _all_ the test
```
> please make a PR changing 

maybe better:
"please bring it up in the babblesim discord channel, or create an RFC issue or PR"
(no need to demand somebody starts by doing all the work, they can start asking or just proposing the change)

jori-nordic replied:
yeah ofc

jori-nordic replied:
but also everybody has grand ideas and I'd like to avoid endless design-by-committee if possible.

aescolar commented on doc/develop/test/bsim.rst:
```diff
@@ -150,23 +158,33 @@ required, and some Zephyr users do not use or have west, but still use the build
+If you have a good idea for improving your test script, please make a PR changing _all_ the test
+scripts in order to benefit everyone and conserve homogeneity.
+
+The exception is scripts starting with an underscore (``_``). They are ignored. They can serve as
```
> The exception is scripts starting with an underscore (``_``). 

I would not make it an exception overall. Having these rot is not going to be too nice. I would still recommend developers to keep them close to the standard ones.

> They are ignored.

I would not phrase it that way.They are just not automatically picked by the batch runner in CI. They may still be used by other scripts.

aescolar commented on doc/develop/test/bsim.rst:
```diff
@@ -150,23 +158,33 @@ required, and some Zephyr users do not use or have west, but still use the build
+
+- Each test is defined by a shell script with the extension ``.sh``, in a subfolder called
+  ``test_scripts/``.
+- It is recommended to run a single test per script file.
```
Nit: Maybe you can mention something like: ". This allows the batcher to better paralelize tests and run more optimally in CI."

aescolar commented on doc/develop/test/bsim.rst:
```diff
@@ -150,23 +158,33 @@ required, and some Zephyr users do not use or have west, but still use the build
+- Tests that require several consecutive simulations (e.g, if simulating a device pairing, powering
+  off, and powering up after as a new simulation) should use separate simulation ids for each
+  simulation segment, ensuring that the radio activity of each segment can be inspected a
   posteriori.
```
I'd ask for three more:
 - Avoid overly long tests. If you tests takes more than a 20 seconds in CI, consider if it is possible to split it in several separate tests
 - If a test takes long to execute (takes more than 5 seconds in CI) be sure to set EXECUTE_TIMEOUT to a value that is at least 5 times bigger than the time it typically takes to run your test in CI. Do not set EXECUTE_TIMEOUT to a value that is lower or equal to its default.
 - Tests should not be overly verbose. A test should not produce more than a few tenths of lines of output to `stdout` and `stderr`. Debug messages are better left disabled.

jori-nordic replied:
The next step in my doc journey is writing a literate-programming (kind of) bsim test where I explain everything in comments. Both in the scripts and in the test code, what libraries to use, conventions etc.
I was thinking of adding the early-exit thing and considerations about unnecessary runtime there.

jori-nordic replied:
To be fair, the last point is because GH's UI sucks. It just chokes on the output if it's more than a few hundred lines. It shouldn't.

aescolar reviewed (changes requested):
Thanks @jori-nordic a few minor things + a few more points I think we should have

jori-nordic committed:
doc: bsim: Add three more rules

On the request of the maintainer.

Background:

We have recently had issues with tests timing out because of the
difference of speed between CI runners and a standard developer machine.

In that case, the logs are printed, and some tests (by yours truly) are
a bit on the chatty side. Github's UI chokes (for a good minute or more
lol) on more than a few hundred lines, so it's a bad experience for
contributors.

Signed-off-by: Jonathan Rico <jonathan.rico@nordicsemi.no>

aescolar commented on doc/develop/test/bsim.rst:
```diff
@@ -150,23 +150,41 @@ required, and some Zephyr users do not use or have west, but still use the build
+variables, etc..
+
+If you have a good idea for improving your test script, please make a PR changing _all_ the test
+scripts in order to benefit everyone and conserve homogeneity. You can of course discuss it first in an RFC issue or on the babblesim discord channel.
```
Silly nit: I'm surprised the compliance check did not complain about this line length..

kartben replied:
Because there isn't any for .rst files 😶

aescolar commented on doc/develop/test/bsim.rst:
```diff
@@ -150,23 +150,41 @@ required, and some Zephyr users do not use or have west, but still use the build
+  off, and powering up after as a new simulation) should use separate simulation ids for each
+  simulation segment, ensuring that the radio activity of each segment can be inspected a
   posteriori.
+- Avoid overly long tests. If the test takes >20 seconds of CPU time, consider if it is possible to
```
Maybe instead of "CPU time", "runtime". (in this context, it doesnt matter if the test is burning the CPU or idling waiting for something else)

aescolar reviewed (changes requested):
Just a minor thing, otherwise lgtm

kartben commented on doc/develop/test/bsim.rst:
```diff
@@ -150,23 +150,42 @@ required, and some Zephyr users do not use or have west, but still use the build
+build system or compatibility changes if the tests are run in the same manner, with the same
+variables, etc..
+
+If you have a good idea for improving your test script, please make a PR changing _all_ the test
```
```suggestion
If you have a good idea for improving your test script, please make a PR changing *all* the test
```

kartben commented on doc/develop/test/bsim.rst:
```diff
@@ -150,23 +150,42 @@ required, and some Zephyr users do not use or have west, but still use the build
-  for the simulated devices and physical layer simulation with the necessary command line options.
-- Tests must return 0 to the invoking shell if the test passes, and not 0 if the test fails.
-- It is recommended to have a single test for each test script.
+Please follow the conventions and not design one-off bespoke runners (e.g. a python script, or
```
```suggestion
Please follow the existing conventions and do not design one-off bespoke runners (e.g. a python script, or
```

kartben commented on doc/develop/test/bsim.rst:
```diff
@@ -150,23 +150,42 @@ required, and some Zephyr users do not use or have west, but still use the build
+  off, and powering up after as a new simulation) should use separate simulation ids for each
+  simulation segment, ensuring that the radio activity of each segment can be inspected a
   posteriori.
+- Avoid overly long tests. If the test takes >20 seconds of runtime, consider if it is possible to
+  split it in several separate tests.
+- If the test takes >5 seconds, set ``EXECUTE_TIMEOUT`` to a value that is at least 5x bigger than
+  the measured run-time.
```
"over 20 seconds" / "over 5 seconds" / "5 times"
